repository/src/utils.rs

```diff
@@ -0,0 +1,100 @@
+use super::*;
+use thiserror::Error;
+
+pub async fn add_remotes<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    known_peers: &[Peer],
+) -> Result<(), Error> {
+    for peer in known_peers {
+        let remote_name = peer.name.clone();
+        let remote_url = format!("git://{}/repo", peer.address.ip());
+        if let Err(err) = this.raw.add_remote(remote_name, remote_url.clone()).await {
+            warn!("failed to add remote({}): {}", remote_url, err);
+        }
+    }
+    for (i, mirror) in this.config.mirrors.iter().enumerate() {
+        let remote_name = format!("mirror_{}", i);
+        if let Err(err) = this.raw.add_remote(remote_name, mirror.clone()).await {
+            warn!("failed to add remote({}): {}", mirror, err);
+        }
+    }
+    Ok(())
+}
+
+/// Retrieve all the finalization proofs
+pub async fn retrieve_fps<T: RawRepository>(
+    this: &DistributedRepository<T>,
+    remote_branches: &[(String, String, CommitHash)],
+) -> Result<Vec<LastFinalizationProof>, Error> {
+    let mut result = Vec::new();
+    for (remote_name, branch_name, commit_hash) in remote_branches {
+        let branch_displayed = format!("{}/{}(at {})", remote_name, branch_name, commit_hash);
+
+        // Skip if the branch is not `fp`.
+        if branch_name != FP_BRANCH_NAME {
+            continue;
+        }
+
+        let fp = this.raw.read_semantic_commit(*commit_hash).await?;
+        let fp = match format::fp_from_semantic_commit(fp) {
+            Ok(x) => x,
+            Err(err) => {
+                warn!("fp branch is invalid {}: {}", branch_displayed, err);
+                continue;
+            }
+        };
+        result.push(fp);
+    }
+    Ok(result)
+}
+
+/// Retrieve all local branches
+pub async fn retrieve_local_branches<T: RawRepository>(
+    raw: &T,
+) -> Result<HashSet<(Branch, CommitHash)>, Error> {
+    let local_branches = raw.list_branches().await?;
+    let mut result = HashSet::new();
+    // TODO: making this concurrent causes a god damn lifetime annoying error
+    for b in local_branches {
+        result.insert((b.clone(), raw.locate_branch(b).await?));
+    }
+    Ok(result)
+}
+
+#[derive(Debug, Error)]
+pub enum CommitError {
+    #[error("raw repo error: {0}")]
+    Raw(#[from] raw::Error),
+    #[error("failed to parse commit ({1}): {0}")]
+    Commit(anyhow::Error, CommitHash),
+}
+
+/// Reads the commits to `Commit`, from the very next commit of ancestor to descendant.
+/// `ancestor` not included, `descendant` included.
+/// It fails if the two commits are the same.
+/// It fails if the ancestor is not the merge base of the two commits.
```

Comment:
> Reads the commits to `Commit`
what do you mean by that?

Reply:
Reads the git commit to the Rust type `Commit` (involving various conversion)

Reply:
oh then `Commit`s please